ProcessLib/Output.cpp

```diff
@@ -0,0 +1,147 @@
+#include<vector>
+
+#include <logog/include/logog.hpp>
+
+#include "BaseLib/FileTools.h"
+
+namespace
+{
+//! Determines if there should be output at the given \c timestep.
+template<typename CountsSteps>
+bool shallDoOutput(std::size_t timestep, CountsSteps const& repeats_each_steps)
+{
+    std::size_t each_steps = 1;
+
+    for (auto const& pair : repeats_each_steps)
+    {
+        each_steps = pair.each_steps;
+
+        if (timestep > pair.repeat * each_steps)
+        {
+            timestep -= pair.repeat * each_steps;
+        } else {
+            break;
+        }
+    }
+
+    return timestep % each_steps == 0;
+}
+}
+
+namespace ProcessLib
+{
+
+template<typename GlobalSetup>
+std::unique_ptr<Output<GlobalSetup>> Output<GlobalSetup>::
+newInstance(const BaseLib::ConfigTree &config, std::string const& output_directory)
+{
+    std::unique_ptr<Output> out{ new Output{
+        BaseLib::joinPaths(output_directory,
+                           config.getConfParam<std::string>("prefix"))}};
+
+    if (auto const timesteps = config.getConfSubtreeOptional("timesteps"))
+    {
+        for (auto pair : timesteps->getConfSubtreeList("pair"))
+        {
+            auto repeat      = pair.getConfParam<std::size_t>("repeat");
+            auto each_steps = pair.getConfParam<std::size_t>("each_steps");
+
+            assert(repeat != 0 && each_steps != 0);
+            out->_repeats_each_steps.emplace_back(repeat, each_steps);
+        }
+
+        if (out->_repeats_each_steps.empty()) {
+            ERR("You have not given any pair (<repeat/>, <each_steps/>) that defines"
+                    " at which timesteps output shall be written. Aborting.");
+            std::abort();
+        }
+    }
+    else
+    {
+        out->_repeats_each_steps.emplace_back(1, 1);
+    }
+
+    return out;
+}
+
+template<typename GlobalSetup>
+void Output<GlobalSetup>::
+init(typename Output<GlobalSetup>::ProcessIter first,
+     typename Output<GlobalSetup>::ProcessIter last)
+{
+    for (unsigned pcs_idx = 0; first != last; ++first, ++pcs_idx)
+    {
+        auto const filename = _output_file_prefix
+                              + "_pcs_" + std::to_string(pcs_idx)
+                              + ".pvd";
+        _single_process_data.emplace(std::piecewise_construct,
+                std::forward_as_tuple(&**first),
+                std::forward_as_tuple(pcs_idx, filename));
+    }
+}
+
+template<typename GlobalSetup>
+void Output<GlobalSetup>::
+doOutputAlways(Process<GlobalSetup> const& process,
+               std::size_t timestep,
+               const double t,
+               typename GlobalSetup::VectorType const& x)
+{
+    auto spd_it = _single_process_data.find(&process);
+    if (spd_it == _single_process_data.end()) {
+        ERR("The given process is not contained in the output configuration."
+            " Aborting.");
+        std::abort();
+    }
+    auto& spd = spd_it->second;
+
+    std::string const output_file_name =
```

Comment:
I think the timestep number should be kept in the output file name since it guarantees unique file names; in contrast `std::to_string(t)` does not (probably it will give unique file names for all practical situations, but what if 6 decimal places don't suffice?).
Regarding your adaptive timestepping argument: Probably there should be different timestep numbers only if we change the adaptivity algorithm; and if we do so, not only the name but also the contents of the files will change. Thus, a change in timestep number would mean a rather fundamental change in our code and there will be manual work anyway, so we can also change the timestep numbers in the filename. Or did you get different timestep numbers with the same configuration on different machines? If so: Might that point to a serious problem, since adaptivity is not stable when changing machines?
Furthermore: The output is currently controlled by timestep number, not by time, so changing the former would also change tha latter. Sure, for adaptivity that will have to change (in a future PR).
To summarize: I think we should keep the timestep number for the time being.